src/environment.py

```diff
@@ -0,0 +1,298 @@
+"""Generate container app environment based on charm configuration."""
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import os
+from urllib.parse import urlparse
+
+# Mattermost's default port, and what we expect the image to use
+CONTAINER_PORT = 8065
+# Default port, enforced via envConfig to prevent operator error
+METRICS_PORT = 8067
+DATABASE_NAME = "mattermost"
+REQUIRED_S3_SETTINGS = ["s3_bucket", "s3_region", "s3_access_key_id", "s3_secret_access_key"]
+REQUIRED_SETTINGS = ["mattermost_image_path"]
+REQUIRED_SSO_SETTINGS = ["license", "site_url"]
+SAML_IDP_CRT = "saml-idp.crt"
+
+
+def _env_for_canonical_defaults(config: dict) -> dict:
```

Comment:
Should this charm have Canonical specific functions in it?

Reply:
I haven't decided that (it existed before). Probably a discussion for a spec